src/MNNLoader.cpp

```diff
@@ -1,132 +1,23 @@
 #include <sstream>
+#include "FileManager.hpp"
 #include "MNNLoader.hpp"
 
-#if 0
 namespace sgns
 {
-
-    MNNLoader::MNNLoader(const std::string &mnn_file_name,
-            unsigned int num_thread) :
-            m_log_id(mnn_file_name.data()), m_mnn_file_path(
-                    mnn_file_name.data()), m_num_threads(num_thread)
-    {
-        initialize();
-    } // End constructor MNNLoader
-
-    MNNLoader::~MNNLoader()
+    SINGLETON_PTR_INIT(MNNLoader);
+    MNNLoader::MNNLoader()
     {
-        // Release model
-        if (m_mnn_interpreter)
-        {
-            m_mnn_interpreter->releaseModel();
-        }
-        // Release all session
-        if (m_mnn_session)
-        {
-            m_mnn_interpreter->releaseSession(m_mnn_session);
-        }
-    } // End ~MNNLoader
+        FileManager::GetInstance().RegisterLoader("file:", this);
+    }
 
-    void MNNLoader::initialize()
+    std::shared_ptr<void> MNNLoader::LoadFile(std::string filename)
     {
+        m_log_id = filename.data();
+        m_mnn_file_path = filename.data();
         // Initialize the interpreter
         m_mnn_interpreter = std::shared_ptr<MNN::Interpreter>(
                 MNN::Interpreter::createFromFile(m_mnn_file_path));
```

Comment:
This shouldn't be just a call to the MNN::Interpreter to load the file how it sees fit.

The biggest issue, is that you should add Async Load/Parse and Save Functions.

Then this MNNLoader should be an async Saver from a void pointer not a loader directly from the file.

Hopefully the MNN::Interpreter, really should act like this.

We need to be able to stream from ipfs: or ftp: or file:  or at least:

LoadAsync("test.mnn", [](void *loadout) { ParseAsync(loadout, [](void *pout) { SaveAsync("mnn", pout) } ) } );

- or -

Pseudo code

shared_ptr<void> LoadAsync(std::string filePath, std::string suffix, shared_ptr<void> *callback());

void ParseFinishedCallback(std::string suffix, shared_ptr<void> asyncOut) {
  SaveAsync(filePath, suffix, asyncOut);  <---- The MNN::Interpreter, will be called to save the data into the interpreter.
}

void LoadFinishedCallback(std::string filePath, std::string suffix, std::string<void> loadout) {
      ParseAsync(filePath, suffix, loadout, ParseFinishedCallback); <---- MNN parse could do nothing, if it doesn't convert from XML/JSON to binary for instance.
}

LoadAsync("file://test.mnn", "mnn", LoadFinishedCallBack);  <--- calls the regular async file load in C++ standard libraray

Reply:
Updated!